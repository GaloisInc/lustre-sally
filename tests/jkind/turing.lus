-- Turing machine simulated in Lustre.
-- 4-state busy beaver on a 2-symbol language:
--
-- http://en.wikipedia.org/wiki/Busy_beaver
--
-- Expected output:
--   INVALID PROPERTY: cex || K = 108
--
--
-- The encoding of the tape uses three integer variables: left, head,
-- and right.
--
--   * The symbol underneath the tape head is stored in 'head'.
--
--   * Everything to the left of the tape head is stored in 'left'
--     using a special base-10 (for readability) encoding. The least
--     significant digit of 'left' is the symbol one cell to the left
--     of the tape head, the second least significant digit of 'left'
--     is the symbol two cells to the left of the tape head, etc.

--   * Everything to the right of the tape head is stored in 'right'
--     using a similar base-10 notation. The least significant digit
--     of 'right' is the symbol one cell to the right of the tape
--     head, the second least significant digit of 'right' is the
--     symbol two cells to the right of the tape head, etc.
--
-- For example, consider the tape:
--
--   ... 0 1 0 1 1 [0] 0 1 1 0 1 0 ...
--
-- Where the '...' parts are all zero and the tape head is at [0].
-- Then left = 1011, head = 0, and right = 10110.
--
-- Since the tape only ever has a finite number of non-zero symbols on
-- it, this encoding is sufficient. Moreover, we easily shift the tape
-- left and right using 'div 10' and 'mod 10'.

type state = enum { Halt, A, B, C, D };
type move = enum { L, R };
type symbol = int;

node main() returns (left, right, head : int);
var
  write : symbol;
  move : move;
  state : state;

  cex : bool;
let
  (left, head, right) = (0, 0, 0) ->
    if pre state = Halt then pre (left, head, right) else
    if move = L then (pre left div 10, pre left mod 10, 10*(pre right) + write)
                else (10*(pre left) + write, pre right mod 10, pre right div 10);

  (write, move, state) = (0, L, A) -> -- Only the initial state matters here
         if pre (state, head) = (A, 0) then (1, R, B)
    else if pre (state, head) = (A, 1) then (1, L, B)
    else if pre (state, head) = (B, 0) then (1, L, A)
    else if pre (state, head) = (B, 1) then (0, L, C)
    else if pre (state, head) = (C, 0) then (1, R, Halt)
    else if pre (state, head) = (C, 1) then (1, L, D)
    else if pre (state, head) = (D, 0) then (1, R, D)
    else if pre (state, head) = (D, 1) then (0, R, A)
    else (0, L, Halt); -- Only the Halt state matters here

  cex = state <> Halt;
  --%PROPERTY cex;
tel;
