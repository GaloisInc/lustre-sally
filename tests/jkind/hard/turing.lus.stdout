Core Lustre
===========

node ()
  returns (left, right, head)
  assumes ()
  shows (cex)
let
  valof_pre_state_30 = pre state
  valof_pre_head_31 = pre head
  valof_pre_state_32 = pre state
  valof_pre_head_33 = pre head
  valof_pre_state_34 = pre state
  valof_pre_head_35 = pre head
  valof_pre_state_36 = pre state
  valof_pre_head_37 = pre head
  valof_pre_state_38 = pre state
  valof_pre_head_39 = pre head
  valof_pre_state_40 = pre state
  valof_pre_head_41 = pre head
  valof_pre_state_42 = pre state
  valof_pre_head_43 = pre head
  valof_pre_state_44 = pre state
  valof_pre_head_45 = pre head
  write = 0 -> ITE(And(Eq(valof_pre_state_30, 1), Eq(valof_pre_head_31, 0)), 1, ITE(And(Eq(valof_pre_state_32, 1), Eq(valof_pre_head_33, 1)), 1, ITE(And(Eq(valof_pre_state_34, 2), Eq(valof_pre_head_35, 0)), 1, ITE(And(Eq(valof_pre_state_36, 2), Eq(valof_pre_head_37, 1)), 0, ITE(And(Eq(valof_pre_state_38, 3), Eq(valof_pre_head_39, 0)), 1, ITE(And(Eq(valof_pre_state_40, 3), Eq(valof_pre_head_41, 1)), 1, ITE(And(Eq(valof_pre_state_42, 4), Eq(valof_pre_head_43, 0)), 1, ITE(And(Eq(valof_pre_state_44, 4), Eq(valof_pre_head_45, 1)), 0, 0))))))))
  valof_pre_state_46 = pre state
  valof_pre_head_47 = pre head
  valof_pre_state_48 = pre state
  valof_pre_head_49 = pre head
  valof_pre_state_50 = pre state
  valof_pre_head_51 = pre head
  valof_pre_state_52 = pre state
  valof_pre_head_53 = pre head
  valof_pre_state_54 = pre state
  valof_pre_head_55 = pre head
  valof_pre_state_56 = pre state
  valof_pre_head_57 = pre head
  valof_pre_state_58 = pre state
  valof_pre_head_59 = pre head
  valof_pre_state_60 = pre state
  valof_pre_head_61 = pre head
  move = 0 -> ITE(And(Eq(valof_pre_state_46, 1), Eq(valof_pre_head_47, 0)), 1, ITE(And(Eq(valof_pre_state_48, 1), Eq(valof_pre_head_49, 1)), 0, ITE(And(Eq(valof_pre_state_50, 2), Eq(valof_pre_head_51, 0)), 0, ITE(And(Eq(valof_pre_state_52, 2), Eq(valof_pre_head_53, 1)), 0, ITE(And(Eq(valof_pre_state_54, 3), Eq(valof_pre_head_55, 0)), 1, ITE(And(Eq(valof_pre_state_56, 3), Eq(valof_pre_head_57, 1)), 0, ITE(And(Eq(valof_pre_state_58, 4), Eq(valof_pre_head_59, 0)), 1, ITE(And(Eq(valof_pre_state_60, 4), Eq(valof_pre_head_61, 1)), 1, 0))))))))
  valof_pre_state_18 = pre state
  valof_pre_left_19 = pre left
  valof_pre_left_20 = pre left
  valof_pre_left_21 = pre left
  left = 0 -> ITE(Eq(valof_pre_state_18, 0), valof_pre_left_19, ITE(Eq(move, 0), Div(valof_pre_left_20, 10), Add(Mul(10, valof_pre_left_21), write)))
  valof_pre_state_26 = pre state
  valof_pre_right_27 = pre right
  valof_pre_right_28 = pre right
  valof_pre_right_29 = pre right
  right = 0 -> ITE(Eq(valof_pre_state_26, 0), valof_pre_right_27, ITE(Eq(move, 0), Add(Mul(10, valof_pre_right_28), write), Div(valof_pre_right_29, 10)))
  valof_pre_state_22 = pre state
  valof_pre_head_23 = pre head
  valof_pre_left_24 = pre left
  valof_pre_right_25 = pre right
  head = 0 -> ITE(Eq(valof_pre_state_22, 0), valof_pre_head_23, ITE(Eq(move, 0), Mod(valof_pre_left_24, 10), Mod(valof_pre_right_25, 10)))
  valof_pre_state_62 = pre state
  valof_pre_head_63 = pre head
  valof_pre_state_64 = pre state
  valof_pre_head_65 = pre head
  valof_pre_state_66 = pre state
  valof_pre_head_67 = pre head
  valof_pre_state_68 = pre state
  valof_pre_head_69 = pre head
  valof_pre_state_70 = pre state
  valof_pre_head_71 = pre head
  valof_pre_state_72 = pre state
  valof_pre_head_73 = pre head
  valof_pre_state_74 = pre state
  valof_pre_head_75 = pre head
  valof_pre_state_76 = pre state
  valof_pre_head_77 = pre head
  state = 1 -> ITE(And(Eq(valof_pre_state_62, 1), Eq(valof_pre_head_63, 0)), 2, ITE(And(Eq(valof_pre_state_64, 1), Eq(valof_pre_head_65, 1)), 2, ITE(And(Eq(valof_pre_state_66, 2), Eq(valof_pre_head_67, 0)), 1, ITE(And(Eq(valof_pre_state_68, 2), Eq(valof_pre_head_69, 1)), 3, ITE(And(Eq(valof_pre_state_70, 3), Eq(valof_pre_head_71, 0)), 0, ITE(And(Eq(valof_pre_state_72, 3), Eq(valof_pre_head_73, 1)), 4, ITE(And(Eq(valof_pre_state_74, 4), Eq(valof_pre_head_75, 0)), 4, ITE(And(Eq(valof_pre_state_76, 4), Eq(valof_pre_head_77, 1)), 1, 0))))))))
  cex = Neq(state, 0)
tel
Sally Model
===========

(define-state-type
   S
   (
     (cex Bool)
     (head Int)
     (left Int)
     (move Int)
     (right Int)
     (state Int)
     (write Int)
     (|gal-initializing| Bool)
     (|head:init| Bool)
     (|left:init| Bool)
     (|move:init| Bool)
     (|right:init| Bool)
     (|state:init| Bool)
     (|valof_pre_head:23| Int)
     (|valof_pre_head:31| Int)
     (|valof_pre_head:33| Int)
     (|valof_pre_head:35| Int)
     (|valof_pre_head:37| Int)
     (|valof_pre_head:39| Int)
     (|valof_pre_head:41| Int)
     (|valof_pre_head:43| Int)
     (|valof_pre_head:45| Int)
     (|valof_pre_head:47| Int)
     (|valof_pre_head:49| Int)
     (|valof_pre_head:51| Int)
     (|valof_pre_head:53| Int)
     (|valof_pre_head:55| Int)
     (|valof_pre_head:57| Int)
     (|valof_pre_head:59| Int)
     (|valof_pre_head:61| Int)
     (|valof_pre_head:63| Int)
     (|valof_pre_head:65| Int)
     (|valof_pre_head:67| Int)
     (|valof_pre_head:69| Int)
     (|valof_pre_head:71| Int)
     (|valof_pre_head:73| Int)
     (|valof_pre_head:75| Int)
     (|valof_pre_head:77| Int)
     (|valof_pre_left:19| Int)
     (|valof_pre_left:20| Int)
     (|valof_pre_left:21| Int)
     (|valof_pre_left:24| Int)
     (|valof_pre_right:25| Int)
     (|valof_pre_right:27| Int)
     (|valof_pre_right:28| Int)
     (|valof_pre_right:29| Int)
     (|valof_pre_state:18| Int)
     (|valof_pre_state:22| Int)
     (|valof_pre_state:26| Int)
     (|valof_pre_state:30| Int)
     (|valof_pre_state:32| Int)
     (|valof_pre_state:34| Int)
     (|valof_pre_state:36| Int)
     (|valof_pre_state:38| Int)
     (|valof_pre_state:40| Int)
     (|valof_pre_state:42| Int)
     (|valof_pre_state:44| Int)
     (|valof_pre_state:46| Int)
     (|valof_pre_state:48| Int)
     (|valof_pre_state:50| Int)
     (|valof_pre_state:52| Int)
     (|valof_pre_state:54| Int)
     (|valof_pre_state:56| Int)
     (|valof_pre_state:58| Int)
     (|valof_pre_state:60| Int)
     (|valof_pre_state:62| Int)
     (|valof_pre_state:64| Int)
     (|valof_pre_state:66| Int)
     (|valof_pre_state:68| Int)
     (|valof_pre_state:70| Int)
     (|valof_pre_state:72| Int)
     (|valof_pre_state:74| Int)
     (|valof_pre_state:76| Int)
     (|write:init| Bool))
   ())
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |write:init| true)
      (= |move:init| true)
      (= |left:init| true)
      (= |right:init| true)
      (= |head:init| true)
      (= |state:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|valof_pre_state:30| state.state)
      (= next.|valof_pre_head:31| state.head)
      (= next.|valof_pre_state:32| state.state)
      (= next.|valof_pre_head:33| state.head)
      (= next.|valof_pre_state:34| state.state)
      (= next.|valof_pre_head:35| state.head)
      (= next.|valof_pre_state:36| state.state)
      (= next.|valof_pre_head:37| state.head)
      (= next.|valof_pre_state:38| state.state)
      (= next.|valof_pre_head:39| state.head)
      (= next.|valof_pre_state:40| state.state)
      (= next.|valof_pre_head:41| state.head)
      (= next.|valof_pre_state:42| state.state)
      (= next.|valof_pre_head:43| state.head)
      (= next.|valof_pre_state:44| state.state)
      (= next.|valof_pre_head:45| state.head)
      (ite
         state.|write:init|
         (= next.write 0)
         (=
            next.write
            (ite
               (and
                  (= next.|valof_pre_state:30| 1)
                  (= next.|valof_pre_head:31| 0))
               1
               (ite
                  (and
                     (= next.|valof_pre_state:32| 1)
                     (= next.|valof_pre_head:33| 1))
                  1
                  (ite
                     (and
                        (= next.|valof_pre_state:34| 2)
                        (= next.|valof_pre_head:35| 0))
                     1
                     (ite
                        (and
                           (= next.|valof_pre_state:36| 2)
                           (= next.|valof_pre_head:37| 1))
                        0
                        (ite
                           (and
                              (= next.|valof_pre_state:38| 3)
                              (= next.|valof_pre_head:39| 0))
                           1
                           (ite
                              (and
                                 (= next.|valof_pre_state:40| 3)
                                 (= next.|valof_pre_head:41| 1))
                              1
                              (ite
                                 (and
                                    (= next.|valof_pre_state:42| 4)
                                    (= next.|valof_pre_head:43| 0))
                                 1
                                 (ite
                                    (and
                                       (= next.|valof_pre_state:44| 4)
                                       (= next.|valof_pre_head:45| 1))
                                    0
                                    0))))))))))
      (= next.|write:init| false)
      (= next.|valof_pre_state:46| state.state)
      (= next.|valof_pre_head:47| state.head)
      (= next.|valof_pre_state:48| state.state)
      (= next.|valof_pre_head:49| state.head)
      (= next.|valof_pre_state:50| state.state)
      (= next.|valof_pre_head:51| state.head)
      (= next.|valof_pre_state:52| state.state)
      (= next.|valof_pre_head:53| state.head)
      (= next.|valof_pre_state:54| state.state)
      (= next.|valof_pre_head:55| state.head)
      (= next.|valof_pre_state:56| state.state)
      (= next.|valof_pre_head:57| state.head)
      (= next.|valof_pre_state:58| state.state)
      (= next.|valof_pre_head:59| state.head)
      (= next.|valof_pre_state:60| state.state)
      (= next.|valof_pre_head:61| state.head)
      (ite
         state.|move:init|
         (= next.move 0)
         (=
            next.move
            (ite
               (and
                  (= next.|valof_pre_state:46| 1)
                  (= next.|valof_pre_head:47| 0))
               1
               (ite
                  (and
                     (= next.|valof_pre_state:48| 1)
                     (= next.|valof_pre_head:49| 1))
                  0
                  (ite
                     (and
                        (= next.|valof_pre_state:50| 2)
                        (= next.|valof_pre_head:51| 0))
                     0
                     (ite
                        (and
                           (= next.|valof_pre_state:52| 2)
                           (= next.|valof_pre_head:53| 1))
                        0
                        (ite
                           (and
                              (= next.|valof_pre_state:54| 3)
                              (= next.|valof_pre_head:55| 0))
                           1
                           (ite
                              (and
                                 (= next.|valof_pre_state:56| 3)
                                 (= next.|valof_pre_head:57| 1))
                              0
                              (ite
                                 (and
                                    (= next.|valof_pre_state:58| 4)
                                    (= next.|valof_pre_head:59| 0))
                                 1
                                 (ite
                                    (and
                                       (= next.|valof_pre_state:60| 4)
                                       (= next.|valof_pre_head:61| 1))
                                    1
                                    0))))))))))
      (= next.|move:init| false)
      (= next.|valof_pre_state:18| state.state)
      (= next.|valof_pre_left:19| state.left)
      (= next.|valof_pre_left:20| state.left)
      (= next.|valof_pre_left:21| state.left)
      (ite
         state.|left:init|
         (= next.left 0)
         (=
            next.left
            (ite
               (= next.|valof_pre_state:18| 0)
               next.|valof_pre_left:19|
               (ite
                  (= next.move 0)
                  (ite
                     (< 10 0)
                     (-
                        (to_int
                           (/
                              (to_real next.|valof_pre_left:20|)
                              (to_real
                                 (- 10)))))
                     (to_int
                        (/
                           (to_real next.|valof_pre_left:20|)
                           (to_real 10))))
                  (+
                     (* 10 next.|valof_pre_left:21|)
                     next.write)))))
      (= next.|left:init| false)
      (= next.|valof_pre_state:26| state.state)
      (= next.|valof_pre_right:27| state.right)
      (= next.|valof_pre_right:28| state.right)
      (= next.|valof_pre_right:29| state.right)
      (ite
         state.|right:init|
         (= next.right 0)
         (=
            next.right
            (ite
               (= next.|valof_pre_state:26| 0)
               next.|valof_pre_right:27|
               (ite
                  (= next.move 0)
                  (+
                     (* 10 next.|valof_pre_right:28|)
                     next.write)
                  (ite
                     (< 10 0)
                     (-
                        (to_int
                           (/
                              (to_real next.|valof_pre_right:29|)
                              (to_real
                                 (- 10)))))
                     (to_int
                        (/
                           (to_real next.|valof_pre_right:29|)
                           (to_real 10))))))))
      (= next.|right:init| false)
      (= next.|valof_pre_state:22| state.state)
      (= next.|valof_pre_head:23| state.head)
      (= next.|valof_pre_left:24| state.left)
      (= next.|valof_pre_right:25| state.right)
      (ite
         state.|head:init|
         (= next.head 0)
         (=
            next.head
            (ite
               (= next.|valof_pre_state:22| 0)
               next.|valof_pre_head:23|
               (ite
                  (= next.move 0)
                  (mod next.|valof_pre_left:24| 10)
                  (mod next.|valof_pre_right:25| 10)))))
      (= next.|head:init| false)
      (= next.|valof_pre_state:62| state.state)
      (= next.|valof_pre_head:63| state.head)
      (= next.|valof_pre_state:64| state.state)
      (= next.|valof_pre_head:65| state.head)
      (= next.|valof_pre_state:66| state.state)
      (= next.|valof_pre_head:67| state.head)
      (= next.|valof_pre_state:68| state.state)
      (= next.|valof_pre_head:69| state.head)
      (= next.|valof_pre_state:70| state.state)
      (= next.|valof_pre_head:71| state.head)
      (= next.|valof_pre_state:72| state.state)
      (= next.|valof_pre_head:73| state.head)
      (= next.|valof_pre_state:74| state.state)
      (= next.|valof_pre_head:75| state.head)
      (= next.|valof_pre_state:76| state.state)
      (= next.|valof_pre_head:77| state.head)
      (ite
         state.|state:init|
         (= next.state 1)
         (=
            next.state
            (ite
               (and
                  (= next.|valof_pre_state:62| 1)
                  (= next.|valof_pre_head:63| 0))
               2
               (ite
                  (and
                     (= next.|valof_pre_state:64| 1)
                     (= next.|valof_pre_head:65| 1))
                  2
                  (ite
                     (and
                        (= next.|valof_pre_state:66| 2)
                        (= next.|valof_pre_head:67| 0))
                     1
                     (ite
                        (and
                           (= next.|valof_pre_state:68| 2)
                           (= next.|valof_pre_head:69| 1))
                        3
                        (ite
                           (and
                              (= next.|valof_pre_state:70| 3)
                              (= next.|valof_pre_head:71| 0))
                           0
                           (ite
                              (and
                                 (= next.|valof_pre_state:72| 3)
                                 (= next.|valof_pre_head:73| 1))
                              4
                              (ite
                                 (and
                                    (= next.|valof_pre_state:74| 4)
                                    (= next.|valof_pre_head:75| 0))
                                 4
                                 (ite
                                    (and
                                       (= next.|valof_pre_state:76| 4)
                                       (= next.|valof_pre_head:77| 1))
                                    1
                                    0))))))))))
      (= next.|state:init| false)
      (=
         next.cex
         (not
            (= next.state 0)))))

(query
   TS
   (or |gal-initializing| cex))

[Lustre] Validating properties:
[Lustre] Property cex... [[1;91mError[0m] Sally error:
/tmp/sallyXXX30965-0.mcmt(292)  : error 3 : , at offset 18
    near [Index: 1081 (Start: -1138256065-Stop: -1138256065) ='(', type<14> Line: 292 LinePos:18]
     : cannot match to any predicted input...
Parse error: /tmp/sallyXXX30965-0.mcmt:292:18: Parse error: /tmp/sallyXXX30965-0.mcmt:292:18: Parse error.

[[1;91mError[0m] ExitFailure 1

